File: apps/docs/src/prework/examples/functional.ts
import type { Reason, Request } from "./procedural";

type Rule = (request: Request) => readonly Reason[];

const equipmentRule: Rule = (request) =>
  request.equipmentChecked ? [] : ["EquipmentNotChecked"];

const flareRule: Rule = (request) =>
  request.flareAlert ? ["FlareAlertActive"] : [];

// 関数を受け取り、組み合わせた関数を返す。
export const combine = (rules: readonly Rule[]): Rule =>
  (request) => rules.flatMap((rule) => rule(request));

export const check = combine([equipmentRule, flareRule]);

File: apps/docs/src/prework/examples/rust/functional.rs
use crate::procedural::{Reason, Request};

pub type Rule = fn(&Request) -> Vec<Reason>;

fn equipment_rule(request: &Request) -> Vec<Reason> {
    if request.equipment_checked {
        vec![]
    } else {
        vec![Reason::EquipmentNotChecked]
    }
}

fn flare_rule(request: &Request) -> Vec<Reason> {
    if request.flare_alert {
        vec![Reason::FlareAlertActive]
    } else {
        vec![]
    }
}

pub fn combine(rules: Vec<Rule>) -> impl Fn(&Request) -> Vec<Reason> {
    move |request| rules.iter().flat_map(|rule| rule(request)).collect()
}

pub fn check(request: &Request) -> Vec<Reason> {
    combine(vec![equipment_rule, flare_rule])(request)
}
